Sources/PubNub/Networking/Routers/FileManagementRouter.swift

```diff
@@ -0,0 +1,538 @@
+    case "MalformedPOSTRequest":
+      return PubNubError(.fileContentLength, additional: [message])
+    default:
+      return PubNubError(.unknown, additional: [code, message])
+    }
+  }
+}
+
+// MARK: - Publish Request Body
+
+/// The message structure used by all SDKs to denote a published message
+struct FilePublishPayload: PubNubFile {
+  /// The additional payload published along with the file
+  var additionalDetails: JSONCodable? {
+    get {
+      return additionalDetailsConcrete
+    }
+    set {
+      additionalDetailsConcrete = newValue?.codableValue
+    }
+  }
+
+  var additionalDetailsConcrete: AnyJSON?
+
+  /// The ID of file on the server
+  var fileId: String
+  /// The name of the file on the server
+  var filename: String
+  /// The size in bytes of the file
+  var size: Int64
+  /// The MIME Type of the file
+  var contentType: String?
+
+  var concreteCustom: AnyJSON?
+  var custom: JSONCodable? {
+    get { return concreteCustom }
+    set { concreteCustom = newValue?.codableValue }
+  }
+
+  var channel: String
+  var createdDate: Date?
+
+  var concreteMessage: AnyJSON?
+  var message: JSONCodable? {
+    get { return concreteCustom }
+    set { concreteCustom = newValue?.codableValue }
+  }
+
+  init(
+    channel: String,
+    fileId: String,
+    filename: String,
+    size: Int64 = 0,
+    contentType: String? = nil,
+    createdDate: Date? = nil,
+    custom: JSONCodable? = nil
+  ) {
+    self.channel = channel
+    self.fileId = fileId
+    self.filename = filename
+    self.size = size
+    self.contentType = contentType ?? "application/octet-stream"
+    self.createdDate = createdDate
+    self.custom = custom
+  }
+
+  init(from other: PubNubFile) {
+    self.init(
+      channel: other.channel,
+      fileId: other.fileId,
+      filename: other.filename,
+      size: other.size,
+      contentType: other.contentType,
+      createdDate: other.createdDate,
+      custom: other.custom
+    )
+  }
+
+  // MARK: Codable
+
+  enum CodingKeys: String, CodingKey {
+    case message
+    case file
+    case channel
+  }
+
+  enum FileCodingKeys: String, CodingKey {
+    case id
+    case name
+    case size
+    case mimeType = "mime-type"
+    case custom
+  }
+
+  init(from decoder: Decoder) throws {
+    let container = try decoder.container(keyedBy: CodingKeys.self)
+    concreteMessage = try container.decodeIfPresent(AnyJSON.self, forKey: .message)
+    channel = try container.decodeIfPresent(String.self, forKey: .channel) ?? ""
+
+    let fileContainer = try container.nestedContainer(keyedBy: FileCodingKeys.self, forKey: .file)
+    fileId = try fileContainer.decode(String.self, forKey: .id)
+    filename = try fileContainer.decode(String.self, forKey: .name)
+    size = try fileContainer.decode(Int64.self, forKey: .size)
+    contentType = try fileContainer.decode(String.self, forKey: .mimeType)
+    concreteCustom = try fileContainer.decodeIfPresent(AnyJSON.self, forKey: .custom)
```

Comment:
It's not on the spec, but the goal was to provide a customization base to the file objects to be more robust.  I was hesitant to include it here, and will move it to the framework layer.

Reply:
@CraigLn let's not forget, that this payload will arrive not only to Swift SDK. If there ever will appear another SDK, which map response to models, it may experience some issue, because Swift will include unexpected field in payload.

Reply:
Agreed, and it has been removed.